File: 0146-lru-cache/0146-lru-cache.cpp
class LRUCache {
public:
    class Node{
        public:
            int key;
            int val;
            Node* next;
            Node* prev;

            Node(int _key, int _val){
                key = _key;
                val = _val;
            }
    };

    Node*head = new Node(-1, -1);
    Node*tail = new Node(-1, -1);
    unordered_map<int, Node*> mapp;
    int cap;

    LRUCache(int capacity) {
        cap = capacity;
        head -> next = tail;
        tail -> prev = head;
    }

    void delNode(Node* resNode){
        Node* prevNode = resNode -> prev;
        Node* nextNode = resNode -> next;

        prevNode -> next = nextNode;
        nextNode -> prev = prevNode;
    }

    void addNode(Node* resNode){
        resNode -> next = head -> next;
        resNode -> prev = head;
        head -> next -> prev = resNode;
        head -> next = resNode;
    }
    
    int get(int key) {
        if (mapp.find(key) != mapp.end()){
            Node* resNode = mapp[key];
            int ans = resNode -> val;
            mapp.erase(key);

            // delete Node 
            delNode(resNode);
            // Insert Node 
            addNode(resNode);
            mapp[key] = head -> next;
            return ans;
        }
        return -1;
    }
    
    void put(int key, int value) {
        if(mapp.find(key) != mapp.end()){
            Node* resNode = mapp[key];
            delNode(resNode);
            mapp.erase(key);
        }
        if(mapp.size()== cap){
            mapp.erase(tail -> prev -> key);
            delNode(tail -> prev);
        }
        addNode(new Node(key, value));
        mapp[key] = head -> next;
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: 0146-lru-cache/0146-lru-cache.py
class LRUCache:
    class Node:
        def __init__(self, key, value):
            self.key = key
            self.value = value
            self.prev = None
            self.next = None

    def __init__(self, capacity: int):
        self.cap = capacity
        self.map = {}    #key -> Node
        self.head = self.Node(-1, -1)
        self.tail = self.Node(-1, -1)
        self.head.next = self.tail
        self.tail.prev = self.head

    # {1:2, 2:3}

    def addNode(self, node):
        currNode = self.head.next
        node.next = currNode
        node.prev = self.head
        currNode.prev = node
        self.head.next = node

    def delNode(self, node):
        prevNode = node.prev
        nextNode = node.next
        prevNode.next = nextNode
        nextNode.prev = prevNode

    def get(self, key: int) -> int:
        if key in self.map.keys():
            node = self.map[key]
            ans = node.value
            self.delNode(node)
            self.addNode(node)
            self.map[key] = self.head.next
            return ans

        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.map.keys():
            node = self.map[key]
            self.delNode(node)
            del self.map[key]
        
        if len(self.map) == self.cap:
            del self.map[self.tail.prev.key]
            self.delNode(self.tail.prev)

        newNode = self.Node(key, value)
        self.addNode(newNode)
        self.map[key] = self.head.next


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)